src/EditorFeatures/Core/Implementation/LineSeparators/LineSeparatorTaggerProvider.cs

```diff
@@ -22,31 +27,53 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.LineSeparators
     [TagType(typeof(LineSeparatorTag))]
     [ContentType(ContentTypeNames.CSharpContentType)]
     [ContentType(ContentTypeNames.VisualBasicContentType)]
-    internal partial class LineSeparatorTaggerProvider :
-        AbstractAsynchronousBufferTaggerProvider<LineSeparatorTag>
+    internal partial class LineSeparatorTaggerProvider : AsynchronousTaggerProvider<LineSeparatorTag>
     {
         [ImportingConstructor]
         public LineSeparatorTaggerProvider(
             IForegroundNotificationService notificationService,
             [ImportMany] IEnumerable<Lazy<IAsynchronousOperationListener, FeatureMetadata>> asyncListeners)
-            : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.LineSeparators), notificationService)
+                : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.LineSeparators), notificationService)
         {
         }
 
-        protected override bool RemoveTagsThatIntersectEdits => true;
-
-        protected override SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
-
-        protected override ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        public override ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
         {
             return TaggerEventSources.Compose(
                 TaggerEventSources.OnTextChanged(subjectBuffer, TaggerDelay.NearImmediate),
                 TaggerEventSources.OnOptionChanged(subjectBuffer, FeatureOnOffOptions.LineSeparator, TaggerDelay.NearImmediate));
         }
 
-        protected override ITagProducer<LineSeparatorTag> CreateTagProducer()
+        public override async Task ProduceTagsAsync(TaggerContext<LineSeparatorTag> context, DocumentSnapshotSpan documentSnapshotSpan, int? caretPosition)
         {
-            return new TagProducer();
+            var cancellationToken = context.CancellationToken;
+            var document = documentSnapshotSpan.Document;
+            if (document == null)
+            {
+                return;
+            }
+
+            var options = document.Project.Solution.Workspace.Options;
+            if (!options.GetOption(FeatureOnOffOptions.LineSeparator, document.Project.Language))
+            {
+                return;
+            }
+
+            // note: we are not directly using the syntax tree root here, we are holding onto it so that the 
+            // line separator service won't block trying to get it.
```

Comment:
This would have been a problem back when we weren't fully async. I'm guessing you can delete the comment? I don't see the syntax tree root anywhere in any case!